// grpl/machinetoleaguenight.js
var Q = require('q');

function getPool(){ return module.parent.exports.pool; }

exports.MachineToLeagueNight = MachineToLeagueNight;
function MachineToLeagueNight(starts){
	this.starts = starts;
	this.machines = {};
}
MachineToLeagueNight.prototype.starts = null;
MachineToLeagueNight.prototype.machines = {};
MachineToLeagueNight.prototype.clear = function(){
	this.machines = {};
}
MachineToLeagueNight.prototype.add = function(abbv, display_order){
	this.machines[abbv] = display_order;
}
MachineToLeagueNight.prototype.save = function(){
	var self = this,
		d = Q.defer();

	getPool().getConnection(function(err, connection){
		if(err){ d.reject(err); return false; }

		var keys = [];

		for(var abbv in self.machines){
			var data = {
				'abbv': abbv,
				'starts': self.starts,
				'display_order': self.machines[abbv]
			};

			keys.push(abbv);

			var query = connection.query("INSERT INTO machine_to_league_night SET ? ON DUPLICATE KEY UPDATE ?", [data, data], function(err, result) {
				if(err){ d.reject(err); return false; }
			});
			// console.log(query.sql);
		}

		
		var del_query = connection.query("DELETE FROM machine_to_league_night WHERE starts=? AND abbv NOT IN (?)", [self.starts, keys], function(err,result){
			if(err){ d.reject(err); return false; }
			connection.end();
			d.resolve(true);
		});
		// console.log(del_query.sql);
	});

	return d.promise;
}
