// grpl/playersublist.js
var Q = require('q');

function getPool(){ return module.parent.exports.pool; }

exports.PlayerSubList = PlayerSubList;
function PlayerSubList(starts){
	this.starts = starts;
	this.subs = [];
}
PlayerSubList.prototype.starts = null;
PlayerSubList.prototype.subs = [];
PlayerSubList.prototype.clear = function(){
	this.subs = [];
}
PlayerSubList.prototype.add = function(data){
	this.subs.push( data );
}
PlayerSubList.prototype.save = function(){
	var self = this,
		d = Q.defer(),
		query_defers = [];

	console.log(self.subs);

	getPool().getConnection(function(err, connection){
		if(err){ d.reject(err); return false; }

		var ids = [];

		for(var i in self.subs){
			if(isNaN(self.subs[i]))
				self.subs[i] == null
			else
				ids.push(self.subs[i])
		}

		var sql = "DELETE FROM league_night_sub WHERE starts=? ";
		if(ids.length > 0)
			sql += "AND sub_id NOT IN (?)";
		else
			sql += "?";

		var del_query = connection.query(sql, [self.starts, ids], function(err, r){
			if(err){ d.reject(err); return false; }

			for(var i in self.subs){
				var data = self.subs[i],
					qd = Q.defer();

				data.starts = self.starts;
				query_defers.push(qd);

				var query = connection.query("INSERT INTO league_night_sub SET ? ON DUPLICATE KEY UPDATE ?", [data, data], function(err, result) {
					if(err){ qd.reject(err); return false; }

					qd.resolve(true);
				});
				// console.log(query.sql);
			}
		});
		// console.log(del_query.sql);

		Q.all(query_defers)
		.then(function(){
			d.resolve(true);
		})
		.fail(function(err){
			d.reject(err);
		})
		.done(function(){
			connection.end();
		});
		
	});

	return d.promise;
}
