// grpl/player.js
var Q = require('q');

/*
 *	Controller Functions
*/
exports.getByNameKey = function(name_key){
	var d = Q.defer();

	getPool().getConnection(function(err, db){
		if(err){ d.reject(err); return false; }
		db.query("SELECT * FROM player WHERE name_key=?", [name_key], function(err, results){
			if(err){ d.reject(err); return false; }
			d.resolve(new Player(results[0]));
			db.end();
		});
	});

	return d.promise;
};

exports.getByFBToken = function(token){
	var https = require('https'),
		d = Q.defer();

	https.get('https://graph.facebook.com/me?fields=id&access_token='+token, function(res) {
		res.on('data', function(data) {
			var fb_id = JSON.parse(data).id;
			exports.getByFBID(fb_id)
			.then(function(player){
				d.resolve(player);
			})
			.fail(function(err){
				d.reject(err);
			});
		});
	}).on('error', function(e) {
		d.reject(e);
	});

	return d.promise;
}


exports.getByFBID = function(fb_id){
	var d = Q.defer();

	getPool().getConnection(function(err, db){
		if(err){ d.reject(err); return false; }
		db.query("SELECT * FROM player WHERE facebook_id=?", [fb_id], function(err, results){
			if(err){ d.reject(err); return false; }
			if(results.length == 0){ 
				d.reject(new Error('could not find user with that ID')); 
				db.end(); 
				return false;
			}
			d.resolve(new Player(results[0]));
			db.end();
		});
	});

	return d.promise;
}

exports.getByHash = function(hash){
	var d = Q.defer();

	getPool().getConnection(function(err, db){
		if(err){ d.reject(err); return false; }
		db.query("SELECT * FROM player WHERE hash=?", [hash], function(err, results){
			if(err){ d.reject(err); return false; }
			d.resolve(new Player(results[0]));
			db.end();
		});
	});

	return d.promise;
}

exports.getByEmailAndPassword = function(email, password){
	var d = Q.defer();

	getPool().getConnection(function(err, db){
		if(err){ d.reject(err); return false; }
		db.query("SELECT * FROM player WHERE email=? AND hash = MD5(CONCAT(name_key,'-',email,'-',?,'-',IFNULL(facebook_id,'')))", [email, password], function(err, results){
			if(err){ d.reject(err); return false; }
			d.resolve(new Player(results[0]));
			db.end();
		});
	});

	return d.promise;
}

/*
 *	Takes a player and saves the email/password needed to be able to login
*/
exports.register = function(data){
	var d = Q.defer();

	getPool().getConnection(function(err, db){
		if(err){ d.reject(err); return false; }
		var sql = "INSERT INTO player SET "+
		"name_key=?, first_name=?, last_name=?, facebook_id=?, email=?, hash=MD5(CONCAT(?,'-',IFNULL(facebook_id,''))) "+
		"ON DUPLICATE KEY UPDATE name_key=VALUES(name_key), first_name=VALUES(first_name), last_name=VALUES(last_name), facebook_id=VALUES(facebook_id), email=VALUES(email), hash=VALUES(hash)";

		db.query(sql, [data.name_key, data.first_name, data.last_name, data.facebook_id, data.email, data.name_key+'-'+data.email+'-'+data.password], function(err, results){
			if(err){ d.reject(err); return false; }
			d.resolve(true);
			db.end();
		});
	});

	return d.promise;
}





function getPool(){ return module.parent.exports.pool; }

/*
 *	Actual Object
*/
exports.Player = Player;
function Player(opts){
	for(prop in opts){
		if (prop in this) {
			this[prop] = opts[prop];
		}
	}

	this.admin = !!this.admin;
	this.initials = this.first_name[0]+this.last_name[0];
}
Player.prototype.player_id = null;
Player.prototype.first_name = null;
Player.prototype.last_name = null;
Player.prototype.initials = null;
Player.prototype.email = null;
Player.prototype.facebook_id = null;
Player.prototype.admin = null;
Player.prototype.name_key = null;
Player.prototype.score = null;
Player.prototype.night_score = null;
Player.prototype.machines = null;
Player.prototype.place = null;
Player.prototype.hash = null;
Player.prototype.getNightTotals = function(season_id){
	var self = this,
		d = Q.defer(),
		sql = "SELECT " + 
			"ln.title, DATE_FORMAT(ln.starts, '%Y-%m-%d') AS starts, pppn.points, lns.sub " +
		"FROM league_night ln " + 
			"LEFT JOIN player_points_per_night pppn USING(starts) " +
			"LEFT JOIN league_night_sub lns USING(starts, name_key) " +
		"WHERE season_id=? " +
			"AND name_key=? " +
		"ORDER BY starts DESC";

	getPool().getConnection(function(err, db){
		if(err){ d.reject(err); return false; }
		var query = db.query(sql, [season_id, self.name_key], function(err, rows){
			if(err){ d.reject(err); return false; }
			d.resolve(rows);
			db.end();
		});
	});

	return d.promise;
}
Player.prototype.getMachinePoints = function(season_id){
	var d = Q.defer(),
		self = this;

	var sql = "SELECT " +
		"m.*, lns.points, DATE_FORMAT(lns.starts, '%Y-%m-%d') AS starts, lnsub.sub " +
	"FROM " +
		"league_night l " +
		"LEFT JOIN league_night_score lns USING(starts) " +
		"LEFT JOIN machine m USING(abbv) " +
		"LEFT JOIN league_night_sub lnsub USING(starts, name_key) " +
	"WHERE " +
		"l.season_id=? " +
		"AND lns.name_key=? " +
	"ORDER BY m.name";

	getPool().getConnection(function(err, db){
		if(err){ d.reject(err); return false; }
		var query = db.query(sql, [season_id, self.name_key], function(err, rows){
			if(err){ d.reject(err); return false; }
			d.resolve(rows);
			db.end();
		});
	});

	return d.promise;

}


Player.prototype.save = function(){
	var self = this;
	getPool().getConnection(function(err, connection){
		var query = connection.query("INSERT INTO player SET ? ON DUPLICATE KEY UPDATE ?", [self, self], function(err, result) {
			console.log(result);
		});
	});
}
