// grpl/season.js
var Q = require('q');

exports.getAll = function(){
	var d = Q.defer();

	getPool().getConnection(function(err, db){
		if(err){ d.reject(err); return false; }
		db.query("SELECT season_id, title FROM season", function(err, results){
			if(err){ d.reject(err); return false; }

			var seasons = [];
			results.forEach(function(row){
				seasons.push(new Season(row.season_id, row.title));
			});
			d.resolve(seasons);
			db.end();
		})
	});

	return d.promise;
}

exports.getById = function(id){
	var d = Q.defer();

	getPool().getConnection(function(err, db){
		if(err){ d.reject(err); return false; }
		db.query("SELECT season_id, title FROM season WHERE season_id=?", [id], function(err, results){
			if(err){ d.reject(err); return false; }
			var row = results[0];
			d.resolve(new Season(row.season_id, row.title));
			db.end();
		})
	});

	return d.promise;
}

function getPool(){
	return module.parent.exports.pool;
}


// this allows us to create new empty items by saying new grpl.season.Season();
exports.Season = Season;

function Season(id, title){
	this.season_id = id;
	this.title = title;
}
Season.prototype.season_id = null;
Season.prototype.title = null;
Season.prototype.save = function(){
	var self = this;
	getPool().getConnection(function(err, connection){
		var query = connection.query("INSERT INTO season SET ? ON DUPLICATE KEY UPDATE ?", [self, self], function(err, result) {
			console.log(result);
		});
	});
}
